---
title: "Differential Expression Analysis"
author: "Funmi Oyebamiji"
date: "2024-06-28"
output:
  html_document:
    keep_md: true
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# ===================
# Load required packages
# ===================
``` r
library(R.utils)
library(DESeq2) 
library(tidyverse) #dplyr, tibble
library(ggrepel)
library(EnhancedVolcano)
library(pheatmap)
library(clusterProfiler)
library(org.Hs.eg.db)
library(msigdbr)
library(ReactomePA)

```

# ===================
# Setup & directories
# ===================
``` r
sampleA <- "Ctrl"
sampleB <- "Trt"

if (!dir.exists("QC_Plots")) dir.create("QC_Plots")
if (!dir.exists("DEfiles")) dir.create("DEfiles")
if (!dir.exists("Pathway")) dir.create("Pathway")

# Fill in metadata
PI_project <- ""  # e.g., "LungRNA"
PI_name <- ""     # e.g., "DrSmith"
PI_genome <- ""   # e.g., "hg38"
Projects_dir <- ""   # e.g., "/home/ubuntu/Projects"
orgdb <- org.Hs.eg.db # e.g., org.Mm.eg.db
#Aligned_date <- "" # e.g., "2024-11-01"
#s3_bucket_used <- "" # e.g., "mybucket"
#Pipeline_name <- "nfcore_rnaseqPipeline_results"

starsalmonCounts_folder <- file.path(Projects_dir, paste0(PI_name, "-", PI_project, "-", PI_genome))
if (!dir.exists(starsalmonCounts_folder)) dir.create(starsalmonCounts_folder, recursive = TRUE)

# Sync gene count data from S3
#starsalmonCounts_in_s3 <- paste0("s3://", s3_bucket_used, "/", PI_name, "/", Pipeline_name, "_", Aligned_date, "/star_salmon/")
#sync_cmd <- paste("sudo aws s3 sync", starsalmonCounts_in_s3, starsalmonCounts_folder,
#                  '--exclude "*" --include "salmon.merged.gene_counts.tsv"')
#system(sync_cmd, intern = TRUE)
```

# ===================
# Load counts and metadata
# ===================
``` r
setwd(starsalmonCounts_folder)
df1 <- read.csv("salmon.merged.gene_counts", header = TRUE, sep = "\t")
df1 <- aggregate(. ~ gene_name, data = df1[,-1], mean)
gene_counts <- df1[,-1]
rownames(gene_counts) <- df1[,1]

# Check column names and build condition vector accordingly
samples <- colnames(gene_counts)
# Assume sample naming are arranged in "Ctrl" then "Trt" order
condition <- factor(c(sampleA, sampleA, sampleB, sampleB))
meta_data <- data.frame(sample = samples,
                        condition = factor(condition),
                        row.names = samples)

```

# ===================
# DESeq2 pipeline
# ===================
``` r
dds_new <- DESeqDataSetFromMatrix(countData = round(gene_counts),
                                  colData = meta_data,
                                  design = ~ condition)
dds <- DESeq(dds_new)
resultsNames(dds)

# Differential expression
res <- results(dds, contrast = c("condition", sampleB, sampleA))
LFC <- lfcShrink(dds, contrast = c("condition", sampleB, sampleA), res = res, type = "ashr")
res_df <- as.data.frame(LFC) %>%
  rownames_to_column(var = "gene") %>%
  mutate(change = ifelse(padj < 0.05, "DE", "Not"),
         direction = case_when(
           change == "DE" & log2FoldChange > 1 ~ "Up",
           change == "DE" & log2FoldChange < -1 ~ "Down",
           TRUE ~ "Not"
         ))

write.csv(res_df, paste0("DEfiles/", sampleB, "vs", sampleA, "_result.csv"), row.names = FALSE)

# Filter significant
LFC.sig <- res_df %>% filter(padj < 0.05 & !is.na(padj)) %>% arrange(desc(abs(log2FoldChange)))
```

# ===================
# Plots
# ===================
``` r
# PCA Plot
dds_vst <- vst(dds, blind = TRUE)
pca_data <- plotPCA(dds_vst, intgroup = "condition", returnData = TRUE)
percentVar <- round(100 * attr(pca_data, "percentVar"))

p <- ggplot(pca_data, aes(PC1, PC2, color = condition, label = rownames(pca_data))) +
  geom_point(size = 3) +
  geom_text_repel() +
  xlab(paste0("PC1: ", percentVar[1], "% variance")) +
  ylab(paste0("PC2: ", percentVar[2], "% variance")) +
  theme_bw()

ggsave(paste0("QC_Plots/PCAplot_", sampleB, "vs", sampleA, ".pdf"), plot = p, width = 13, height = 9)

# Volcano Plot
vol_lab <- EnhancedVolcano(res_df,  
                            lab = res_df$gene, x = 'log2FoldChange', y = 'pvalue', pCutoff = 10e-4, FCcutoff = 2, 
                            xlim = c(-10, 10), ylim = c(0, 200), pointSize = 1.5, labSize = 2.5, title = 'DESeq2 results', 
                            subtitle = 'Differential expression', caption = 'FC cutoff, 2; p-value cutoff, 1e-4', 
                            legendPosition = "right", legendLabSize = 14, col = c('grey30', 'forestgreen', 'royalblue', 'red2'), 
                            colAlpha = 0.9, drawConnectors = TRUE, hline = c(1e-8), widthConnectors = 0.5,selectLab = LFC.sig$gene[1:30])

ggsave(filename = paste0("QC_Plots/Volcanoplot_",sampleB,"vs",sampleA,".pdf"), plot = vol_lab, width=13, height=9)

# Heatmap of DE Genes
counts_vst <- assay(dds_vst)
counts_vst_DE <- counts_vst[rownames(counts_vst) %in% LFC.sig$gene,] # counts_vst_DE <- counts_vst[res$change=="DE",] 
mydata_col <- data.frame(condition = meta_data$condition, row.names = rownames(meta_data)) # row.names = meta_data$sample) 

heat_DE <- pheatmap(counts_vst_DE, scale="row", color = colorRampPalette(c("dark blue", "white", "darkred"))(1000), 
                    annotation_col = mydata_col, 
                    main=paste0(sampleB,"vs",sampleA ," FDR < 0.05"),  show_rownames = F) 
                    
pdf(paste0("QC_Plots/Heatmap_",sampleB,"vs",sampleA,".pdf"), width=13, height=9) 
grid::grid.newpage() 
grid::grid.draw(heat_DE$gtable) 
dev.off()
```


# ===================
# Functional Enrichment Analysis
# ===================
``` r

eg1 <- bitr(res_df$gene, fromType = "SYMBOL", toType = "ENTREZID", OrgDb = orgdb)
eg2 <- merge(res_df, eg1, by = 1) 

# Helper function to save enrichment results and plots
save_enrichment <- function(enrich_obj, filename_prefix) {
  # Convert results to dataframe
  res_df <- as.data.frame(enrich_obj@result)
  
  # Save TXT
  write.table(res_df, paste0(filename_prefix, ".txt"), quote = FALSE, sep = "\t", row.names = FALSE)
  
  # Create and save dotplot
  dp <- dotplot(enrich_obj)
  ggsave(filename = paste0(filename_prefix, ".pdf"), plot = dp, width = 13, height = 9)
}

# Filter significant genes
deg_sig <- eg2 %>% filter(padj < 0.05)

# KEGG enrichment
ekg <- enrichKEGG(gene = deg_sig$ENTREZID, organism = "hsa", pvalueCutoff = 0.1, qvalueCutoff = 0.2) # organism = "mmu"
ekg <- setReadable(ekg, OrgDb = orgdb, keyType = "ENTREZID")
save_enrichment(ekg, paste0("Pathway/", sampleB, "vs", sampleA, ".KEGG"))

# GO enrichment (all ontologies)
go_enrich <- enrichGO(gene = deg_sig$ENTREZID, OrgDb = orgdb, ont = "ALL",
                      pAdjustMethod = "BH", pvalueCutoff = 0.1, qvalueCutoff = 0.2, readable = TRUE)
save_enrichment(go_enrich, paste0("Pathway/", sampleB, "vs", sampleA, ".GO"))

# Reactome pathway enrichment
reactome_enrich <- enrichPathway(gene = deg_sig$ENTREZID, organism = "human",
                                 pAdjustMethod = "BH", pvalueCutoff = 0.1, qvalueCutoff = 0.2, readable = TRUE) # organism = "mouse"
save_enrichment(reactome_enrich, paste0("Pathway/", sampleB, "vs", sampleA, ".Reactome"))

# Prepare gene list for GSEA: named vector of log2 fold changes, sorted decreasingly
gene_list <- res_df %>% 
    arrange(desc(log2FoldChange)) %>% 
    dplyr::select(gene, log2FoldChange) %>% 
    deframe()

# Load MSigDB Hallmark gene sets for mouse
msig_mouse_hallmark <- msigdbr(species = "Homo sapiens", collection = "H") %>%
   dplyr::select(gs_name, gene_symbol) # species = "Mus musculus"

# Run GSEA
gsea_result <- GSEA(geneList = gene_list, TERM2GENE = msig_mouse_hallmark,
                    pvalueCutoff = 0.1, minGSSize = 3)

# Save GSEA results
save_enrichment(gsea_result, paste0("Pathway/", sampleB, "vs", sampleA, ".Hallmark"))


```
